src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/Source.cs

```diff
@@ -0,0 +1,238 @@
+            {
+                return null;
+            }
+
+            var service = document.GetLanguageService<CompletionService>();
+            var description = await service.GetDescriptionAsync(document, roslynItem, cancellationToken).ConfigureAwait(false);
+
+            return new ClassifiedTextElement(description.TaggedParts.Select(p => new ClassifiedTextRun(p.Tag.ToClassificationTypeName(), p.Text)));
+        }
+
+        private static RoslynTrigger GetRoslynTrigger(AsyncCompletionData.CompletionTrigger trigger, SnapshotPoint triggerLocation)
+        {
+            switch (trigger.Reason)
+            {
+                case AsyncCompletionData.CompletionTriggerReason.Invoke:
+                case AsyncCompletionData.CompletionTriggerReason.InvokeAndCommitIfUnique:
+                    return RoslynTrigger.Invoke;
+                case AsyncCompletionData.CompletionTriggerReason.Insertion:
+                    return RoslynTrigger.CreateInsertionTrigger(trigger.Character);
+                case AsyncCompletionData.CompletionTriggerReason.Deletion:
+                    return RoslynTrigger.CreateDeletionTrigger(trigger.Character);
+                case AsyncCompletionData.CompletionTriggerReason.Backspace:
+                    var snapshotBeforeEdit = trigger.ViewSnapshotBeforeTrigger;
+                    char characterRemoved;
+                    if (triggerLocation.Position >= 0 && triggerLocation.Position < snapshotBeforeEdit.Length)
+                    {
+                        // If multiple characters were removed (selection), this finds the first character from the left. 
+                        characterRemoved = snapshotBeforeEdit[triggerLocation.Position];
+                    }
+                    else
+                    {
+                        characterRemoved = (char)0;
+                    }
+
+                    return RoslynTrigger.CreateDeletionTrigger(characterRemoved);
+                case AsyncCompletionData.CompletionTriggerReason.SnippetsMode:
+                    return new RoslynTrigger(CompletionTriggerKind.Snippets);
+            }
+
+            FatalError.ReportWithoutCrash(ExceptionUtilities.UnexpectedValue(trigger.Reason));
+            return RoslynTrigger.Invoke;
+        }
+
+        private VSCompletionItem Convert(
+            Document document,
+            RoslynCompletionItem roslynItem,
+            CompletionService completionService,
+            Dictionary<string, AsyncCompletionData.CompletionFilter> filterCache)
+        {
+            var imageId = roslynItem.Tags.GetFirstGlyph().GetImageId();
+            var filters = GetFilters(roslynItem, filterCache);
+            
+            // roslynItem generated by providers can contain an insertionText in a property bag.
+            // Try using it first.
+            if (!roslynItem.Properties.TryGetValue(InsertionText, out var insertionText))
+            {
+                insertionText = roslynItem.DisplayText;
```

Comment:
did we lose the conversation/comments about this..?

Reply:
I hope. We didn't lose: we will not consume insertionText in our TryCommit. However, if other providers will need our insertionText, this will be our best attempt for them.

---
In reply to: [222522689](https://github.com/dotnet/roslyn/pull/29016#discussion_r222522689) [](ancestors = 222522689)